### PR TITLE
filx(SharedLibs): Update GCL interceptors

### PR DESCRIPTION
**Pull request by stchar**
This PR is an attempt to continue the work stated in #111 as the author is not able to work on it.

Plan is:
1. Rebase previously started work
2. Comment out previous changes which intend to fix the issue
3. Add tests to reproduce the issue
4. Apply fix again if issue was reproduced, merge only tests if it wasn't

### CODE REVIEW

File: src/main/groovy/com/lesfurets/jenkins/unit/global/lib/LibraryLoader.groovy [hunk 1]
```diff
@@ -111,6 +111,7 @@ class LibraryLoader {
                 }
             }
             record.definedGlobalVars = globalVars
+            println record.definedGlobalVars.collect({k,v->"$k::$v"})
```
Comment (stchar) [inline src/main/groovy/com/lesfurets/jenkins/unit/global/lib/LibraryLoader.groovy]: delete debug output

File: src/test/groovy/com/lesfurets/jenkins/TestInterceptingGCL.groovy [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    String sharedLibOne = this.class
+        .getResource('/libs/test_cross_vars_usage')
+        .getFile()
+
+    String sharedLibTwo = this.class
+        .getResource('/libs/test_cross_class_usage')
+        .getFile()
+
+    @Test
+    void test_vars_interop_library_loaded_with_implicity() throws Exception {
```
Comment (stchar) [inline src/test/groovy/com/lesfurets/jenkins/TestInterceptingGCL.groovy]: Add comments for each tests what it does and why

**Comment by stchar** [comment]
It verifies #104. You can create library class and call pipeline methods without passing it trough pipeline reference.

File: src/test/groovy/com/lesfurets/jenkins/TestInterceptingGCL.groovy [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        runScript('job/library/cross_class/test_dynamic.jenkins')
+        printCallStack()
+        assertCallStack().contains("""ClassA.methodA()""")
+        assertCallStack().contains("""ClassA.sh(echo 'ClassA: I'm field of A')""")
+        assertCallStack().contains("""ClassB.methodB()""")
+        assertCallStack().contains("""ClassB.sh(echo 'ClassB: I'm field of B')""")
+    }
+
+    @Test
+    void test_cross_lib_interop_no_implicity_annotation() throws Exception {
```
Comment (stchar) [inline src/test/groovy/com/lesfurets/jenkins/TestInterceptingGCL.groovy:225]: Cross lib tests are incorrect. 
scenarios should be like:
1. load libA, libB
2. Test pipeline can call vars from the both libs
3. Test libB can call LibA vars
4. Test case if both LibA and LibB have a vars/method with the same signature (who wins?) 
4.1 Verify how it works in real jenins   

Test the same (if possible) with classes (what if you call a static Method). 
Verify the behaviour is the same in real Jenkins instance

File: src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy [hunk 4]
```diff
@@ -24,19 +24,54 @@ class InterceptingGCL extends GroovyClassLoader {
...
         Class clazz = super.parseClass(codeSource, shouldCacheSource)
-        interceptClassMethods(clazz.metaClass, helper)
+        interceptClassMethods(clazz.metaClass, helper, binding)
         return clazz
     }
+
+    @Override
+    Class<?> loadClass(String name) throws ClassNotFoundException {
+        // Source from: groovy-all-2.4.6-sources.jar!/groovy/lang/GroovyClassLoader.java:710
+        println "Loading class: $name"
```
Comment (stchar) [inline src/main/groovy/com/lesfurets/jenkins/unit/InterceptingGCL.groovy]: delete debug output

File: src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy [hunk 5]
```diff
@@ -140,7 +164,18 @@ abstract class BasePipelineTest {
                 updateBuildStatus('UNSTABLE')
             }
         })
+        helper.registerAllowedMethod("withCredentials", [Map, Closure])
         helper.registerAllowedMethod("withCredentials", [List, Closure], withCredentialsInterceptor)
+        helper.registerAllowedMethod('withEnv', [List, Closure], { List list, Closure c ->
+            list.each {
+                //def env = helper.get
+                def item = it.split('=')
+                assert item.size() == 2, "withEnv list does not look right: ${list.toString()}"
```
Comment (sams-gleb) [inline src/main/groovy/com/lesfurets/jenkins/unit/BasePipelineTest.groovy:175]: it's possible to have additional '=' inside the item, e.g. when passing GRADLE_OPTS. Does it make sense to have this validation in the common library?

**Pull request merged**